### ISSUE TITLE
New Feature - Better names in tracing logs

### ISSUE THREAD

**Issue by geffzhang** [ThreeMammals/Ocelot#227]
I built a Ocelot & butterfly environment and found the name of the event unfriendly, referring to this tracing data，Access Http://111.230.101.173/Home/About generate trace data, See tracing data 
http://111.230.101.173:9618/tracing/traces 
Example： MiddlewareFinished: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; 200

2018-02-13 18:15:31.216
Server Receive

2018-02-13 18:15:31.216
AspNetCore BeginRequest

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: App.Metrics.AspNetCore.Tracking.Middleware.ActiveRequestCounterEndpointMiddleware; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.MiddlewareAnalysis.AnalysisMiddleware; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: App.Metrics.AspNetCore.Tracking.Middleware.ErrorRequestMeterMiddleware; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.MiddlewareAnalysis.AnalysisMiddleware; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: App.Metrics.AspNetCore.Tracking.Middleware.PostAndPutRequestSizeHistogramMiddleware; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.MiddlewareAnalysis.AnalysisMiddleware; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: App.Metrics.AspNetCore.Tracking.Middleware.RequestTimerMiddleware; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.MiddlewareAnalysis.AnalysisMiddleware; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: App.Metrics.AspNetCore.Tracking.Middleware.PerRequestTimerMiddleware; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.MiddlewareAnalysis.AnalysisMiddleware; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; /api/values

2018-02-13 18:15:31.216
MiddlewareStarting: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Microsoft.AspNetCore.Builder.Extensions.MapMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.Errors.Middleware.ExceptionHandlerMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.Responder.Middleware.ResponderMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.DownstreamRouteFinder.Middleware.DownstreamRouteFinderMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.Headers.Middleware.HttpHeadersTransformationMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.Request.Middleware.DownstreamRequestInitialiserMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.RateLimit.Middleware.ClientRateLimitMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.RequestId.Middleware.ReRouteRequestIdMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.Authentication.Middleware.AuthenticationMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.Claims.Middleware.ClaimsBuilderMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.Authorisation.Middleware.AuthorisationMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.Headers.Middleware.HttpRequestHeadersBuilderMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.QueryStrings.Middleware.QueryStringBuilderMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.LoadBalancer.Middleware.LoadBalancingMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.DownstreamUrlCreator.Middleware.DownstreamUrlCreatorMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.Cache.Middleware.OutputCacheMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.Request.Middleware.HttpRequestBuilderMiddleware; /api/values

2018-02-13 18:15:31.217
MiddlewareStarting: Ocelot.Requester.Middleware.HttpRequesterMiddleware; /api/values

2018-02-13 18:15:31.350
MiddlewareFinished: Ocelot.Requester.Middleware.HttpRequesterMiddleware; 200

2018-02-13 18:15:31.350
MiddlewareFinished: Ocelot.Request.Middleware.HttpRequestBuilderMiddleware; 200

2018-02-13 18:15:31.350
MiddlewareFinished: Ocelot.Cache.Middleware.OutputCacheMiddleware; 200

2018-02-13 18:15:31.350
MiddlewareFinished: Ocelot.DownstreamUrlCreator.Middleware.DownstreamUrlCreatorMiddleware; 200

2018-02-13 18:15:31.350
MiddlewareFinished: Ocelot.LoadBalancer.Middleware.LoadBalancingMiddleware; 200

2018-02-13 18:15:31.350
MiddlewareFinished: Ocelot.QueryStrings.Middleware.QueryStringBuilderMiddleware; 200

2018-02-13 18:15:31.350
MiddlewareFinished: Ocelot.Headers.Middleware.HttpRequestHeadersBuilderMiddleware; 200

2018-02-13 18:15:31.350
MiddlewareFinished: Ocelot.Authorisation.Middleware.AuthorisationMiddleware; 200

2018-02-13 18:15:31.350
MiddlewareFinished: Ocelot.Claims.Middleware.ClaimsBuilderMiddleware; 200

2018-02-13 18:15:31.350
MiddlewareFinished: Ocelot.Authentication.Middleware.AuthenticationMiddleware; 200

2018-02-13 18:15:31.350
MiddlewareFinished: Ocelot.RequestId.Middleware.ReRouteRequestIdMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Ocelot.RateLimit.Middleware.ClientRateLimitMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Ocelot.Request.Middleware.DownstreamRequestInitialiserMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Ocelot.Headers.Middleware.HttpHeadersTransformationMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Ocelot.DownstreamRouteFinder.Middleware.DownstreamRouteFinderMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Ocelot.Responder.Middleware.ResponderMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Ocelot.Errors.Middleware.ExceptionHandlerMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.Builder.Extensions.MapMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.MiddlewareAnalysis.AnalysisMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: App.Metrics.AspNetCore.Tracking.Middleware.PerRequestTimerMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.MiddlewareAnalysis.AnalysisMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: App.Metrics.AspNetCore.Tracking.Middleware.RequestTimerMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.MiddlewareAnalysis.AnalysisMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: App.Metrics.AspNetCore.Tracking.Middleware.PostAndPutRequestSizeHistogramMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.MiddlewareAnalysis.AnalysisMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: App.Metrics.AspNetCore.Tracking.Middleware.ErrorRequestMeterMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.MiddlewareAnalysis.AnalysisMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: App.Metrics.AspNetCore.Tracking.Middleware.ActiveRequestCounterEndpointMiddleware; 200

2018-02-13 18:15:31.351
MiddlewareFinished: Microsoft.AspNetCore.Builder.UseWhenExtensions+<>c__DisplayClass0_1; 200

2018-02-13 18:15:31.351
AspNetCore EndRequest

2018-02-13 18:15:31.351
Server Send

**Comment by TomPallister** [comment]
@geffzhang How do you think it could be improved?

I would like to test and see if it traces exceptions in the Ocelot middleware!

**Comment by TomPallister** [comment]
After refactoring ocelot for downstream request aggregation the diagnostic listener no longer records any events other than entering and exiting Ocelot. This is because the middleware is no longer asp.net middleware. In this issue we also need to make something listen to the new Ocelot middlewares being called.

**Comment by geffzhang** [comment]
https://andrewlock.net/understanding-your-middleware-pipeline-with-the-middleware-analysis-package/#addingacustomnameforanonymousmiddleware 
app.Properties["analysis.NextMiddlewareName"] = "NanoFabric.Ocelot";  Only one of them can be modified

**Comment by TomPallister** [comment]
@geffzhang I have fixed the issue where tracing stopped working and cleaned up trace names. They now look like this:

```
2018-03-29 07:51:26.968
MiddlewareStarting: Microsoft.AspNetCore.Builder.Extensions.MapMiddleware; /posts/1

2018-03-29 07:51:26.969
MiddlewareStarting: TransitionToOcelotMiddleware; /posts/1

2018-03-29 07:51:26.972
Ocelot.MiddlewareStarted: ExceptionHandlerMiddleware; /posts/1

2018-03-29 07:51:26.979
Ocelot.MiddlewareStarted: ResponderMiddleware; /posts/1

2018-03-29 07:51:26.982
Ocelot.MiddlewareStarted: DownstreamRouteFinderMiddleware; /posts/1

2018-03-29 07:51:26.999
Ocelot.MiddlewareStarted: HttpHeadersTransformationMiddleware; /posts/1

2018-03-29 07:51:27.001
Ocelot.MiddlewareStarted: DownstreamRequestInitialiserMiddleware; /posts/1

2018-03-29 07:51:27.012
Ocelot.MiddlewareStarted: ClientRateLimitMiddleware; /posts/1

2018-03-29 07:51:27.015
Ocelot.MiddlewareStarted: ReRouteRequestIdMiddleware; /posts/1

2018-03-29 07:51:27.018
Ocelot.MiddlewareStarted: AuthenticationMiddleware; /posts/1

2018-03-29 07:51:27.021
Ocelot.MiddlewareStarted: ClaimsBuilderMiddleware; /posts/1

2018-03-29 07:51:27.022
Ocelot.MiddlewareStarted: AuthorisationMiddleware; /posts/1

2018-03-29 07:51:27.025
Ocelot.MiddlewareStarted: HttpRequestHeadersBuilderMiddleware; /posts/1

2018-03-29 07:51:27.027
Ocelot.MiddlewareStarted: QueryStringBuilderMiddleware; /posts/1

2018-03-29 07:51:27.028
Ocelot.MiddlewareStarted: LoadBalancingMiddleware; /posts/1

2018-03-29 07:51:27.036
Ocelot.MiddlewareStarted: DownstreamUrlCreatorMiddleware; /posts/1

2018-03-29 07:51:27.038
Ocelot.MiddlewareStarted: OutputCacheMiddleware; /posts/1

2018-03-29 07:51:27.041
Ocelot.MiddlewareStarted: HttpRequesterMiddleware; /posts/1

2018-03-29 07:51:27.132
OcelotMiddlewareFinished: HttpRequesterMiddleware; /posts/1

2018-03-29 07:51:27.133
OcelotMiddlewareFinished: OutputCacheMiddleware; /posts/1

2018-03-29 07:51:27.133
OcelotMiddlewareFinished: DownstreamUrlCreatorMiddleware; /posts/1

2018-03-29 07:51:27.134
OcelotMiddlewareFinished: LoadBalancingMiddleware; /posts/1

2018-03-29 07:51:27.134
OcelotMiddlewareFinished: QueryStringBuilderMiddleware; /posts/1

2018-03-29 07:51:27.134
OcelotMiddlewareFinished: HttpRequestHeadersBuilderMiddleware; /posts/1

2018-03-29 07:51:27.134
OcelotMiddlewareFinished: AuthorisationMiddleware; /posts/1

2018-03-29 07:51:27.134
OcelotMiddlewareFinished: ClaimsBuilderMiddleware; /posts/1

2018-03-29 07:51:27.134
OcelotMiddlewareFinished: AuthenticationMiddleware; /posts/1

2018-03-29 07:51:27.135
OcelotMiddlewareFinished: ReRouteRequestIdMiddleware; /posts/1

2018-03-29 07:51:27.135
OcelotMiddlewareFinished: ClientRateLimitMiddleware; /posts/1

2018-03-29 07:51:27.135
OcelotMiddlewareFinished: DownstreamRequestInitialiserMiddleware; /posts/1

2018-03-29 07:51:27.135
OcelotMiddlewareFinished: HttpHeadersTransformationMiddleware; /posts/1

2018-03-29 07:51:27.136
OcelotMiddlewareFinished: DownstreamRouteFinderMiddleware; /posts/1

2018-03-29 07:51:27.136
OcelotMiddlewareFinished: ResponderMiddleware; /posts/1

2018-03-29 07:51:27.137
OcelotMiddlewareFinished: ExceptionHandlerMiddleware; /posts/1

2018-03-29 07:51:27.391
MiddlewareFinished: TransitionToOcelotMiddleware; 200

2018-03-29 07:51:27.392
MiddlewareFinished: Microsoft.AspNetCore.Builder.Extensions.MapMiddleware; 200

2018-03-29 07:51:27.394
AspNetCore EndRequest

2018-03-29 07:51:27.394
Server Send
```

What do you think? Is that OK?

**Comment by geffzhang** [comment]
That's great

**Comment by TomPallister** [comment]
@geffzhang cool! Is there anything else that you would like to see in the trace? I guess we could put everything into the log but that might be expensive in terms of serialisation but I don't really know to be honest!

**Comment by geffzhang** [comment]
It's really expensive to put everything in. As a distributed tracking, that's clear.

**Comment by TomPallister** [comment]
OK just released these changes in 5.3.1

I hope that it all works OK! Let me know if there are any problems.
